echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/PipelineConfigsPollingJob.groovy

```diff
@@ -0,0 +1,186 @@
+    } finally {
+      long elapsedMillis = System.currentTimeMillis() - start
+      log.info("Done polling for pipeline configs in ${elapsedMillis/1000}s")
+      registry.timer("pipelineConfigsPollingAgent.executionTimeMillis").record(elapsedMillis, TimeUnit.MILLISECONDS)
+    }
+  }
+
+  /**
+   * Remove all triggers no longer needed
+   * e.g. triggers that are still registered in our scheduler but no longer found on any pipeline
+   * @param pipelineTriggers all active pipeline CRON triggers
+   */
+  void removeStaleTriggers(TriggerRepository pipelineTriggers) {
+    int removed
+    int failed
+
+    try {
+      Set<TriggerKey> triggerKeys = scheduler.getTriggerKeys(
+        GroupMatcher.triggerGroupEquals(Scheduler.DEFAULT_GROUP))
+
+      triggerKeys.each { triggerKey ->
+        if (pipelineTriggers.getTrigger(triggerKey.name) == null) {
+          try {
+            scheduler.unscheduleJob(triggerKey)
+            removed++
+          } catch (Exception e) {
+            log.error("Failed to unschedule job with triggerId: ${triggerKey.name}", e)
+            failed++
+          }
+        }
+      }
+    }
+    catch (Exception e) {
+      log.error("Failed during stale trigger removal", e)
+    }
+
+    if ((removed > 0) || (failed > 0)) {
+      log.debug("Removed $removed stale triggers successfully, $failed removals failed")
+    }
+  }
+
+  /**
+   * Update triggers: Compare triggers in the system with those of the parsed pipelines
+   * create triggers for new pipeline triggers
+   * update triggers that have changed
+   * @param pipelineTriggers all active pipeline CRON triggers
+   */
+  void updateChangedTriggers(TriggerRepository pipelineTriggers) {
+    int addCount
+    int updateCount
+    int failCount
+
+    try {
+      pipelineTriggers.triggers().each { pipelineTrigger ->
+        CronTrigger trigger = scheduler.getTrigger(TriggerKey.triggerKey(pipelineTrigger.id)) as CronTrigger
+
+        if (!trigger) {
+          if (registerNewTrigger(pipelineTrigger)) {
+            addCount++
+          } else {
+            failCount++
+          }
+        } else {
+          if (!TriggerConverter.isInSync(trigger, pipelineTrigger, timeZoneId)) {
+            try {
+              def newTrigger = TriggerConverter.toQuartzTrigger(pipelineTrigger, timeZoneId)
+              scheduler.rescheduleJob(trigger.key, newTrigger)
+              updateCount++
+            } catch (Exception e) {
+              log.error("Failed to update existing trigger: id: ${pipelineTrigger.id} for pipeline: ${pipelineTrigger.parent.id}", e)
+              failCount++
+            }
+          }
+        }
+      }
+    } catch (Exception e) {
+      log.error("Failed to upsert a trigger", e)
+    }
+
+    if ((addCount + updateCount + failCount) > 0) {
+      log.debug("Added $addCount new triggers, updated $updateCount existing triggers, $failCount failed")
+    }
+
+    registry.gauge("trigger.failed_update_count").set(failCount)
```

Comment:
Convention is to use camelCase for metric names (dot delimiters also good).

Reply:
gotcha. Btw, i wanted to ask you if `gauge` is the right thing here